### DISCUSSION TITLE
How to generate a network with a specified porosity and permeability?

### DISCUSSION THREAD

**Discussion by jgostick** [PMEAL/OpenPNM#1824] [Q&A]
I am creating this discussion to answer a question posted by @fb638 on a [different thread](https://github.com/PMEAL/OpenPNM/discussions/1807#discussioncomment-324744) that deserved its own thread.

@fb638 asked:

> Hi, I am new to OpenPNM and I want to generate a nonuniform porous network with a specific porosity as well as permeability. I have tried to do it with porespy but in this case, pores are not connected. I will be extremely grateful if you help me.
>
> thanks in advance

**Comment by jgostick** [comment]
Generating a pore network with specific properties is a common task, but it's a bit of a pain to do.  There is an example notebook [here](https://github.com/PMEAL/OpenPNM/blob/dev/examples/notebooks/geometry/adjusting_pore_size_distributions.ipynb) that explains how to adjust the pore and throat size distribution of a network.  This example assumed that you'll be doing it by hand.  It is possible to automate things using various optimization techniques, but this is quite technical and it might be faster to just do it by hand.  Examples of automated approaches are a [recent paper](https://link.springer.com/article/10.1007/s11242-018-1087-1) by Matthews' group, and another one by [Sharqawy](https://doi.org/10.1016/j.advwatres.2016.10.023).  I did [one during my PhD](https://doi.org/10.1016/j.jpowsour.2007.04.059) manually, which may also be useful for seeing the process.

**Reply by fb638** [reply]
Thank you for your help.

**Comment by xu-kai-xu** [comment]
@fb638, hi, sorry for the delayed reply. I did generate a PNM with random pore size distribution and pore locations. The general idea is the following:

1. choose the pore size distribution you want to use, e.g., normal distribution, Weibull distribution, or log-normal distribution, last two of which is the most frequently used. jeff's PhD work used Weibull distribution. The pore size generation code is like the following one:

```python

from scipy.stats import weibull_min

def weibull_gene():
    # generation pore size distribution pore size
    c = 1.7 # weibull distribution parameters
    r1 = weibull_min.rvs(c=c, size=1000)
    dis = r1.max() - r1.min()
    rmin = r1.min()

    # scale the min and max to what you need
    r1 = 2.5e-5 + (r1-rmin) / dis * 2.5e-5  
    return(r1)
```

2. decide the pore locations. I use [`Delaunay network`](https://pmeal.github.io/OpenPNM/modules/network/delaunay.html?highlight=delaunay#openpnm.network.Delaunay) to generate random pore locations. But now, the pores are not connected. There are no throats yet.

3. find the connected pores and generate throats. I use [`find_neighbor_throats`](https://pmeal.github.io/OpenPNM/search.html?q=find_neighbor_throats) to do that. Then the pore network structure is finished. You just need to add the geometry, phases, physics, and algorithm object to the network. This method can produce a random pore coordination number and a random network structure. But the coordination number is usually too large.

4. Then you just need to tune the pore size range to change the porosity until you are happy with it.

5. I suggest you read the [documentation](https://pmeal.github.io/OpenPNM/index.html) first, which is the best way to know OpenPNM.

**Reply by Devin-one** [reply]
Hello, sorry to disturb you. I am a beginner, and now I am studying the pore size distribution. I want to know how to call these functions. It would be better if you could send me your complete code. Thank you in advance

**Reply by xu-kai-xu** [reply]
hi, @Liqiwen123, you can read this tutorial first: https://github.com/PMEAL/OpenPNM/tree/dev/examples/getting_started
and these are more examples : https://github.com/PMEAL/OpenPNM/tree/dev/examples

If you would like to generate diy PDF, you can try [this](https://github.com/xu-kai-xu/supplementary_for_PNM_simulation/blob/master/article/fig_3_random_probability_density_functions.ipynb). I wrote it for my research work.

**Reply by nltthinh** [reply]
Hi, sorry but another newbie here. I have a bimodal lognormal function from which I would like to generate pore size distribution. Is there any way to define my own function and still be able to call .rsv on that?
Many thanks in advance!
